specifications/sep-payment-channel-mechanism.md

```diff
@@ -0,0 +1,655 @@
+    thresholds such that I and R's signers must both sign, and adding I's
+    signers to E.
+    - One or more `SET_OPTIONS` operations adding I's signers to V.
+    - One or more `CHANGE_TRUST` operations adding trustlines to E.
+    - One `END_SPONSORING_FUTURE_RESERVES` operation that stops I sponsoring
+    future reserves of subsequent operations.
+  - Operations sponsored by R:
+    - One `BEGIN_SPONSORING_FUTURE_RESERVES` operation that specifies reserve
+    account R as a sponsor of future reserves.
+    - One or more `SET_OPTIONS` operations adjusting escrow account V's
+    thresholds such that R and I's signers must both sign, and adding R's
+    signers to V.
+    - One or more `SET_OPTIONS` operations adding R's signers to E.
+    - One `END_SPONSORING_FUTURE_RESERVES` operation that stops R sponsoring
+    future reserves of subsequent operations.
+  - One or more `PAYMENT` operations depositing I's contribution to E.
+  - One or more `PAYMENT` operations depositing R's contribution to E.
+  - One or more `PAYMENT` operations depositing R's reserves to V, for each
+  trustline on E that will be used to sponsor claimable balances at
+  disbursement.
+  
+- C_i, see [Update](#Update) process.
+
+- D_i, see [Update](#Update) process.
+
+#### Update
+
+To update the payment channel state, the participants:
+
+1. Increment i.
+2. Sign and exchange a closing transaction C_i.
+3. Sign and exchange a declaration transaction D_i.
+
+The transactions are constructed as follows:
+
+- C_i, the _closing transaction_, disburses funds to R and changes the signing
+weights on E such that I unilaterally controls E.  C_i has source account E,
+sequence number s_i+1, a `minSeqAge` of O (the observation period time
+duration), and a `minSeqLedgerGap` of O (the observation period ledger count).
+
+  The `minSeqAge` and `minSeqLedgerGap` prevents a misbehaving party from
+  executing C_i when the channel state has already progressed to a later
+  iteration number, as the other party has the period O to invalidate C_i by
+  submitting D_i' for some i' > i.
+  
+  C_i contains operations:
+  - One `BEGIN_SPONSORING_FUTURE_RESERVES` operation that specifies reserve
+  account V as a sponsor of future reserves.
+  - One `CREATE_CLAIMABLE_BALANCE` operation for each trustline that is
+  disbursing funds to R.
+  - One `END_SPONSORING_FUTURE_RESERVES` operation that confirms reserve
+  account V's sponsorship.
+  - One or more `SET_OPTIONS` operation adjusting escrow account E's
+  thresholds to give I full control of E, and removing R's signers.
+  - One or more `SET_OPTIONS` operation adjusting reserve account E's
+  thresholds to give R full control of V, and removing I's signers.
+
+- D_i, the _declaration transaction_, declares an intent to execute the
+corresponding closing transaction C_i.  D_i has source account E, sequence
+number s_i, and `minSeqNum` set to s_e.  Hence, D_i can execute at any time, so
```

Comment:
s_e is a little confusing to me. Say we start at this state:

```
s (E's seqNum) = 1000, i = 0, e = 0
s_i = 1000
s_e = 1000
```

First C_i/D_i get created but not submitted, i =1.
Second C_i/D_i get created and submitted, i = 2, e=1

This second D_i set E.seqNum to 1000 + 2*1 = 1002
This second C_i then sets E.seqNum to 1003

The current state is then:
```
s = 1003, i=2, e=1
s_i = 1004
s_e = 1002
```

So for a third D_i, shouldn't the minSeqNum be 1003, and thus E's current seqNum. And not s_e?

Reply:
We chatted offline about this, so I think this is resolved. The things that we noted are:
 - Once D is submitted, for e.g. D_2 in the example above, the participants don't sign a third transaction set because the channel is closing.
 - s_e doesn't really matter for channel close, so the doc doesn't mention it being updated in the close steps. We could add it for completeness, but I think it's probably unnecessary? If the second D above was submitted, theoretically e would be equal to 2, the iteration number for the transaction set being submitted.

We used the following table to help think about how s_i and s_e change:
|Description                           |Transaction|Tx Seq|E's seq change|E's seq|s   |i change|i  |e change|e  |s_i |s_e |m  |
|--------------------------------------|-----------|------|--------------|-------|----|--------|---|--------|---|----|----|---|
|Setup: 1. I creates escrow account E  |           |      |1000          |1000   |    |        |   |        |   |    |    |2  |
|Setup: 3. Set variable initial states |           |      |              |1000   |1001|0       |0  |0       |0  |1001|1001|2  |
|                                      |           |      |              |1000   |1001|        |0  |        |0  |1001|1001|2  |
|Setup: 5. Build formation transaction.|F          |1001  |              |1000   |1001|        |0  |        |0  |1001|1001|2  |
|                                      |           |      |              |1000   |1001|        |0  |        |0  |1001|1001|2  |
|Setup: 6. Increment i                 |           |      |              |1000   |1001|1       |1  |        |0  |1003|1001|2  |
|Setup: 7. Build and sign C_i          |C_1        |1004  |              |1000   |1001|        |1  |        |0  |1003|1001|2  |
|Setup: 8. Build and sign D_i          |D_1        |1003  |              |1000   |1001|        |1  |        |0  |1003|1001|2  |
|                                      |           |      |              |1000   |1001|        |1  |        |0  |1003|1001|2  |
|Setup: 10. Submit F                   |           |      |1001          |1001   |1001|        |1  |        |0  |1003|1001|2  |
|                                      |           |      |              |1001   |1001|        |1  |        |0  |1003|1001|2  |
|------                                |           |      |              |1001   |1001|        |1  |        |0  |1003|1001|2  |
|                                      |           |      |              |1001   |1001|        |1  |        |0  |1003|1001|2  |
|Update: Increment i                   |           |      |              |1001   |1001|2       |2  |        |0  |1005|1001|2  |
|Update: Build and sign C_i            |C_2        |1006  |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|Update: Build and sign D_i            |D_2        |1005  |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|                                      |           |      |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|------                                |           |      |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|                                      |           |      |              |1001   |1001|        |2  |        |0  |1005|1001|2  |
|Update: Increment i                   |           |      |              |1001   |1001|3       |3  |        |0  |1007|1001|2  |
|Update: Build and sign C_i            |C_3        |1008  |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|Update: Build and sign D_i            |D_3        |1007  |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|                                      |           |      |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|---                                   |           |      |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|                                      |           |      |              |1001   |1001|        |3  |        |0  |1007|1001|2  |
|Close: Submit D_2                     |           |      |1005          |1005   |1001|        |3  |2       |2  |1007|1005|2  |
|Close: Submit D_3                     |           |      |1007          |1007   |1001|        |3  |3       |3  |1007|1007|2  |
|Close: Submit C_3                     |           |      |1008          |1008   |1001|        |3  |        |3  |1007|1007|2  |
|                                      |           |      |              |1008   |1001|        |3  |        |3  |1007|1007|2  |

Ref: https://docs.google.com/spreadsheets/d/1cOXL3mYdMTNDu0FE8lgNMa0fjJ2T-S2jolLG5Dl0xRM

Reply:
yup that answered it thx. Also in case someone else sees this I was wondering about s_e after the close in case _I_ wanted to continue the channel (I was thinking i and e would remain at 3). But after close the variables reset to 0, so s_e after close doesn't matter (as you said)

Reply:
That's a great point. This is an edge case we should make clearer. I added a section about it in ab21927.